chainer/functions/normalization/group_normalization.py

```diff
@@ -1,10 +1,360 @@
+        self.retain_inputs((0,))
+        x, = inputs
+        x_hat = cuda.elementwise(
+            'T x, T mean, T inv_std', 'T x_hat',
+            'x_hat = (x - mean) * inv_std',
+            'groupnorm_x_hat')(x, self.mean[:, None], self.inv_std[:, None])
+        self.retain_outputs((0,))
+        return x_hat,
+
+    def backward(self, indexes, grad_outputs):
+        x, = self.get_retained_inputs()
+        x_hat, = self.get_retained_outputs()
+        gx_hat, = grad_outputs
+        return _XHatGrad(
+            self.eps, self.mean, self.inv_std,
+            self.dummy_gamma, x_hat.array).apply((x, gx_hat))
+
+
+class _XHatGrad(function_node.FunctionNode):
+
+    def __init__(self, eps, mean, inv_std, dummy_gamma, x_hat):
+        self.eps = eps
+        self.mean = mean
+        self.inv_std = inv_std
+        self.dummy_gamma = dummy_gamma
+        self.x_hat = x_hat
+
+    def forward(self, inputs):
+        xp = backend.get_array_module(*inputs)
+        if xp is not numpy and chainer.should_use_cudnn('>=auto', 5000) and \
+                self.dummy_gamma is not None:
+            return self.forward_cudnn(inputs)
+
+        self.retain_inputs((0, 1))
+        _, gx_hat = inputs
+        x_hat = self.x_hat
+        self.x_hat = None
+
+        gx_hat_avg = gx_hat.mean(axis=1, keepdims=True)
+        gx_hat_x_hat_avg = (gx_hat * x_hat).mean(axis=1, keepdims=True)
+        gx_std = gx_hat - gx_hat_avg - x_hat * gx_hat_x_hat_avg
+        gx = self.inv_std[:, None] * gx_std
+
+        self.retain_outputs((0,))
+        return gx,
+
+    def forward_cudnn(self, inputs):
+        if self.eps < libcudnn.CUDNN_BN_MIN_EPSILON:
+            raise RuntimeError(
+                'cuDNN does not allow an eps value '
+                'less than {}.'.format(libcudnn.CUDNN_BN_MIN_EPSILON))
+
+        self.retain_inputs((0, 1))
+        x, gx_hat = inputs
+        self.x_hat = None
+
+        # `x[None, :, :, None]` is slower because it results in a different
+        # strides and cuDNN doesn't recognize it as a contiguous array.
+        reduced_shape = x.shape
+        cudnn_shape = (1,) + reduced_shape + (1,)
+        x = x.reshape(cudnn_shape)
+        gx_hat = gx_hat.reshape(cudnn_shape)
+
+        gx, _, _ = cudnn.batch_normalization_backward(
+            x, self.dummy_gamma, gx_hat,
+            self.mean, self.inv_std, self.eps,
+            True, libcudnn.CUDNN_BATCHNORM_SPATIAL,
+            configuration.config.debug)
+        gx = gx.reshape(reduced_shape)
+
+        self.retain_outputs((0,))
+        return gx,
+
+    def backward(self, indexes, grad_outputs):
+        F = chainer.functions
+        x, gx_hat = self.get_retained_inputs()
+        gx, = self.get_retained_outputs()
+        ggx, = grad_outputs
+
+        x_hat, = _XHat(
+            self.eps, self.mean, self.inv_std,
+            self.dummy_gamma).apply((x,))
+
+        ret = []
+
+        if 0 in indexes:
+            # gx = inv_std * gx_std
```

Comment:
I think these commented out lines can be removed.

Reply:
Those lines are *not* commented out source codes. They are here to sketch the derivation of double backprop formula.

I'd like to make the comments less confusing, but I don't have an idea...

Reply:
apology.

How about adding a single comment that clarifies below lines are not comment, but the sketch of double backprop formula?
For me, just adding some comments at the beginning and end of the sketch is enough.